### ISSUE TITLE
Trading

### ISSUE THREAD

**Issue by chibuikeiheanacho873-droid** [chibuikeiheanacho873-droid/Emmi#1]
import React, { useState, useEffect } from "react"; // Trading Chart Analyzer // Single-file React component (Default export) ready to drop in a React app. // - Uses Tailwind CSS for styling // - Uses react-chartjs-2 + chart.js for charting (install these as deps) // - Uses papaparse to parse CSV upload (install as dependency) // - Implements SMA, EMA, RSI, MACD calculations in JS (no external TA libs) // Features: // 1. Upload CSV (date,open,high,low,close,volume) OR paste JSON array of OHLC. // 2. Visualize close price with overlayed indicators (SMA/EMA/MACD/RSI shown separately). // 3. Run scanner that combines signals (MA crossover, RSI, MACD) into a single Buy/Hold/Sell suggestion with a confidence score. // 4. Option to enter an API key + symbol (AlphaVantage) to fetch data (optional, user-provided key). If you don't want to fetch live data, upload CSV. // 5. Clear comments and explanation of each signal. // How to use: install dependencies: react, react-dom, tailwindcss, chart.js, react-chartjs-2, papaparse // Example: npm i react-chartjs-2 chart.js papaparse

import { Line } from "react-chartjs-2"; import Papa from "papaparse"; import { Chart as ChartJS, CategoryScale, LinearScale, PointElement, LineElement, Title, Tooltip, Legend, } from "chart.js";

ChartJS.register( CategoryScale, LinearScale, PointElement, LineElement, Title, Tooltip, Legend );

// --- Technical indicator implementations (all arrays assume chronological order oldest->newest) function SMA(values, period) { const out = new Array(values.length).fill(null); let sum = 0; for (let i = 0; i < values.length; i++) { sum += values[i]; if (i >= period) sum -= values[i - period]; if (i >= period - 1) out[i] = sum / period; } return out; }

function EMA(values, period) { const out = new Array(values.length).fill(null); const k = 2 / (period + 1); let prev = null; for (let i = 0; i < values.length; i++) { const v = values[i]; if (i === period - 1) { // seed with SMA const seed = values.slice(0, period).reduce((a, b) => a + b, 0) / period; prev = seed; out[i] = seed; } else if (i >= period) { prev = v * k + prev * (1 - k); out[i] = prev; } } return out; }

function RSI(values, period = 14) { const out = new Array(values.length).fill(null); let gain = 0; let loss = 0; for (let i = 1; i < values.length; i++) { const change = values[i] - values[i - 1]; gain += Math.max(0, change); loss += Math.max(0, -change); if (i === period) { const avgGain = gain / period; const avgLoss = loss / period; const rs = avgLoss === 0 ? 100 : avgGain / avgLoss; out[i] = 100 - 100 / (1 + rs); var prevAvgGain = avgGain; var prevAvgLoss = avgLoss; } else if (i > period) { const change2 = values[i] - values[i - 1]; const g = Math.max(0, change2); const l = Math.max(0, -change2); prevAvgGain = (prevAvgGain * (period - 1) + g) / period; prevAvgLoss = (prevAvgLoss * (period - 1) + l) / period; const rs2 = prevAvgLoss === 0 ? 100 : prevAvgGain / prevAvgLoss; out[i] = 100 - 100 / (1 + rs2); } } return out; }

function MACD(values, fast = 12, slow = 26, signal = 9) { const emaFast = EMA(values, fast); const emaSlow = EMA(values, slow); const macd = new Array(values.length).fill(null); for (let i = 0; i < values.length; i++) { if (emaFast[i] !== null && emaSlow[i] !== null) macd[i] = emaFast[i] - emaSlow[i]; } const signalLine = EMA(macd.map((v) => (v === null ? 0 : v)), signal); // seed zeros for nulls const hist = macd.map((v, i) => (v === null || signalLine[i] === null ? null : v - signalLine[i])); return { macd, signalLine, hist }; }

// Helper to parse CSV rows to objects function normalizeRows(rows) { // expects columns: date,open,high,low,close,volume (case-insensitive). If headers missing, tries to guess. const headers = Object.keys(rows[0] || {}); if (headers.length >= 5) { return rows.map((r) => ({ date: r.date || r.Date || r.time || r.Time || r.Datetime || r.datetime, open: parseFloat(r.open || r.Open), high: parseFloat(r.high || r.High), low: parseFloat(r.low || r.Low), close: parseFloat(r.close || r.Close), volume: parseFloat(r.volume || r.Volume) || 0, })).filter(x => !Number.isNaN(x.close)); } return []; }

export default function TradingAnalyzer() { const [rawRows, setRawRows] = useState([]); const [symbol, setSymbol] = useState(""); const [indicators, setIndicators] = useState(null); const [decision, setDecision] = useState(null); const [apiKey, setApiKey] = useState(""); const [status, setStatus] = useState("");

// parse CSV upload function handleFile(e) { const file = e.target.files[0]; if (!file) return; Papa.parse(file, { header: true, dynamicTyping: false, complete: (results) => { const rows = normalizeRows(results.data); rows.sort((a, b) => new Date(a.date) - new Date(b.date)); setRawRows(rows); setStatus(Loaded ${rows.length} rows from CSV); }, }); }

// parse pasted JSON function handlePasteJson() { const text = prompt("Paste JSON array of {date,open,high,low,close,volume}"); if (!text) return; try { const arr = JSON.parse(text); const rows = normalizeRows(arr); rows.sort((a, b) => new Date(a.date) - new Date(b.date)); setRawRows(rows); setStatus(Loaded ${rows.length} rows from pasted JSON); } catch (e) { alert("Invalid JSON"); } }

// Optional: fetch from AlphaVantage (user supplies API key). THIS IS OPTIONAL and may hit rate limits. We keep it simple. async function fetchAlphaVantage(sym) { if (!apiKey) { alert("Provide your AlphaVantage API key (optional)"); return; } try { setStatus("Fetching data..."); const url = https://www.alphavantage.co/query?function=TIME_SERIES_DAILY_ADJUSTED&symbol=${encodeURIComponent(sym)}&outputsize=compact&apikey=${apiKey}; const res = await fetch(url); const data = await res.json(); if (data['Error Message'] || !data['Time Series (Daily)']) throw new Error("No data"); const ts = data['Time Series (Daily)']; const rows = Object.keys(ts).map((d) => ({ date: d, open: parseFloat(ts[d]['1. open']), high: parseFloat(ts[d]['2. high']), low: parseFloat(ts[d]['3. low']), close: parseFloat(ts[d]['4. close']), volume: parseFloat(ts[d]['6. volume'] || ts[d]['5. volume'] || 0), })); rows.sort((a, b) => new Date(a.date) - new Date(b.date)); setRawRows(rows); setStatus(Fetched ${rows.length} rows for ${sym}); } catch (e) { console.error(e); setStatus("Failed to fetch. Check symbol/api key or rate limits."); } }

// compute indicators when rawRows changes useEffect(() => { if (!rawRows || rawRows.length < 10) { setIndicators(null); setDecision(null); return; } const closes = rawRows.map((r) => r.close); const sma20 = SMA(closes, 20); const sma50 = SMA(closes, 50); const ema12 = EMA(closes, 12); const ema26 = EMA(closes, 26); const rsi14 = RSI(closes, 14); const macdObj = MACD(closes, 12, 26, 9);

const ind = { sma20, sma50, ema12, ema26, rsi14, macd: macdObj };
setIndicators(ind);
setDecision(computeDecision(rawRows, ind));

}, [rawRows]);

// scanning / decision logic function computeDecision(rows, ind) { // Evaluate latest signals using last available non-null values const lastIndex = rows.length - 1; function last(arr) { for (let i = lastIndex; i >= 0; i--) if (arr[i] !== null && arr[i] !== undefined) return { value: arr[i], index: i }; return { value: null, index: -1 }; } const close = rows[lastIndex].close; const sma20 = last(ind.sma20).value; const sma50 = last(ind.sma50).value; const ema12 = last(ind.ema12).value; const ema26 = last(ind.ema26).value; const rsi = last(ind.rsi14).value; const macd = last(ind.macd.macd).value; const macdSignal = last(ind.macd.signalLine).value; const macdHist = last(ind.macd.hist).value;

const signals = [];

// Moving average crossover (20 vs 50)
if (sma20 && sma50) {
  if (sma20 > sma50) signals.push({ name: "MA Crossover (20>50)", side: "buy", weight: 1.2, note: "Short-term MA above long-term MA" });
  else if (sma20 < sma50) signals.push({ name: "MA Crossover (20<50)", side: "sell", weight: 1.0, note: "Short-term MA below long-term MA" });
  else signals.push({ name: "MA Crossover (flat)", side: "neutral", weight: 0.1 });
}

// EMA crossover (12 vs 26)
if (ema12 && ema26) {
  if (ema12 > ema26) signals.push({ name: "EMA Crossover (12>26)", side: "buy", weight: 1.1, note: "Momentum rising" });
  else if (ema12 < ema26) signals.push({ name: "EMA Crossover (12<26)", side: "sell", weight: 0.9, note: "Momentum falling" });
}

// RSI
if (rsi !== null && rsi !== undefined) {
  if (rsi < 30) signals.push({ name: "RSI Oversold", side: "buy", weight: 1.3, note: `RSI ${rsi.toFixed(1)} < 30` });
  else if (rsi > 70) signals.push({ name: "RSI Overbought", side: "sell", weight: 1.3, note: `RSI ${rsi.toFixed(1)} > 70` });
  else signals.push({ name: "RSI Neutral", side: "neutral", weight: 0.2, note: `RSI ${rsi.toFixed(1)}` });
}

// MACD histogram momentum
if (macdHist !== null && macdHist !== undefined) {
  if (macdHist > 0) signals.push({ name: "MACD Histogram > 0", side: "buy", weight: 1.0, note: `MACD hist ${macdHist.toFixed(4)}` });
  else if (macdHist < 0) signals.push({ name: "MACD Histogram < 0", side: "sell", weight: 1.0, note: `MACD hist ${macdHist.toFixed(4)}` });
}

// Price vs SMA50
if (sma50) {
  if (close > sma50) signals.push({ name: "Price above SMA50", side: "buy", weight: 0.6 });
  else signals.push({ name: "Price below SMA50", side: "sell", weight: 0.6 });
}

// Aggregate signals into score
let score = 0; // positive => buy, negative => sell
let totWeight = 0;
signals.forEach((s) => {
  const w = s.weight || 1;
  totWeight += Math.abs(w);
  if (s.side === "buy") score += w;
  else if (s.side === "sell") score -= w;
});
const normalized = totWeight === 0 ? 0 : score / totWeight; // -1 .. 1-ish
let recommendation = "HOLD";
if (normalized > 0.15) recommendation = "BUY";
else if (normalized < -0.15) recommendation = "SELL";

const confidence = Math.min(100, Math.round(Math.abs(normalized) * 100));

return { recommendation, confidence, normalized, signals, details: { close, sma20, sma50, ema12, ema26, rsi, macd, macdSignal, macdHist } };

}

// Chart dataset builder function buildChartData() { if (!rawRows || !indicators) return null; const labels = rawRows.map((r) => r.date); const closes = rawRows.map((r) => r.close); const datasets = [ { label: "Close", data: closes, fill: false, tension: 0.1, borderWidth: 2 }, ]; // include SMA/EMA (only if present) if (indicators.sma20) datasets.push({ label: "SMA20", data: indicators.sma20, fill: false, borderDash: [6, 4] }); if (indicators.sma50) datasets.push({ label: "SMA50", data: indicators.sma50, fill: false, borderDash: [6, 4] }); if (indicators.ema12) datasets.push({ label: "EMA12", data: indicators.ema12, fill: false });

return { labels, datasets };

}

const chartData = buildChartData();

return ( <div className="p-6 bg-slate-50 min-h-screen"> <div className="max-w-5xl mx-auto bg-white rounded-2xl shadow-lg p-6"> <h1 className="text-2xl font-semibold mb-2">Trading Chart Analyzer</h1> <p className="text-sm text-slate-600 mb-4">Upload OHLC CSV or paste data, then run the scanner to get a Buy / Hold / Sell suggestion with confidence and signal breakdown.</p>

<div className="grid grid-cols-1 md:grid-cols-3 gap-3 mb-4">
      <div>
        <label className="block text-xs text-slate-500">Upload CSV (date,open,high,low,close,volume)</label>
        <input type="file" accept=".csv" onChange={handleFile} className="mt-1 p-2 border rounded w-full" />
      </div>
      <div>
        <label className="block text-xs text-slate-500">Paste JSON</label>
        <button onClick={handlePasteJson} className="mt-1 w-full px-3 py-2 bg-slate-100 rounded">Paste JSON array</button>
      </div>
      <div>
        <label className="block text-xs text-slate-500">Optional: Fetch symbol (AlphaVantage)</label>
        <input placeholder="SYMBOL e.g. AAPL" value={symbol} onChange={(e)=>setSymbol(e.target.value)} className="mt-1 p-2 border rounded w-full" />
        <input placeholder="AlphaVantage API key (optional)" value={apiKey} onChange={(e)=>setApiKey(e.target.value)} className="mt-1 p-2 border rounded w-full" />
        <div className="flex gap-2 mt-2">
          <button onClick={()=>fetchAlphaVantage(symbol)} className="px-3 py-2 bg-blue-600 text-white rounded">Fetch</button>
          <button onClick={()=>{ setRawRows([]); setIndicators(null); setDecision(null); setStatus('Cleared'); }} className="px-3 py-2 bg-rose-100 rounded">Clear</button>
        </div>
      </div>
    </div>

    <div className="mb-4">
      <div className="flex items-center justify-between">
        <div className="text-sm text-slate-500">Status: {status || 'idle'}</div>
        <div>
          <button onClick={()=>{ if (!rawRows || rawRows.length<10) return alert('Load at least 10 rows'); const d = computeDecision(rawRows, indicators); setDecision(d); setStatus('Scanned'); }} className="px-3 py-2 bg-green-600 text-white rounded">Run Scanner</button>
        </div>
      </div>
    </div>

    <div className="grid grid-cols-1 md:grid-cols-3 gap-4">
      <div className="md:col-span-2">
        <div className="bg-white p-4 rounded shadow-sm">
          {chartData ? (
            <Line data={chartData} options={{ responsive: true, plugins: { legend: { position: 'top' } } }} />
          ) : (
            <div className="text-sm text-slate-500">Upload data to see chart</div>
          )}
        </div>

        {/* RSI and MACD mini panels */}
        {indicators && (
          <div className="grid grid-cols-1 md:grid-cols-2 gap-4 mt-4">
            <div className="bg-white p-4 rounded shadow-sm">
              <h3 className="font-medium">RSI (14) — latest: {indicators && indicators.rsi14 && indicators.rsi14.filter(v=>v!==null).slice(-1)[0] ? indicators.rsi14.filter(v=>v!==null).slice(-1)[0].toFixed(2) : 'n/a'}</h3>
              <div className="text-sm text-slate-500 mt-2">Values are calculated on close price. RSI &lt; 30 is considered oversold (buy), &gt; 70 is overbought (sell).</div>
            </div>
            <div className="bg-white p-4 rounded shadow-sm">
              <h3 className="font-medium">MACD — latest hist: {indicators && indicators.macd && indicators.macd.hist && indicators.macd.hist.filter(v=>v!==null).slice(-1)[0] ? indicators.macd.hist.filter(v=>v!==null).slice(-1)[0].toFixed(6) : 'n/a'}</h3>
              <div className="text-sm text-slate-500 mt-2">MACD histogram > 0 implies bullish momentum.</div>
            </div>
          </div>
        )}
      </div>

      <div>
        <div className="bg-white p-4 rounded shadow-sm">
          <h3 className="text-lg font-semibold">Scanner Result</h3>
          {!decision && <div className="text-sm text-slate-500 mt-2">No scan yet — upload data and run the scanner.</div>}
          {decision && (
            <div className="mt-2">
              <div className={`inline-block px-3 py-1 rounded-full font-medium ${decision.recommendation === 'BUY' ? 'bg-green-100 text-green-800' : decision.recommendation === 'SELL' ? 'bg-rose-100 text-rose-800' : 'bg-slate-100 text-slate-800'}`}>{decision.recommendation}</div>
              <div className="mt-2 text-sm text-slate-600">Confidence: <strong>{decision.confidence}%</strong></div>

              <div className="mt-3">
                <h4 className="font-medium">Signal breakdown</h4>
                <ul className="mt-2 space-y-2 text-sm">
                  {decision.signals.map((s, i) => (
                    <li key={i} className="p-2 border rounded">
                      <div className="flex justify-between">
                        <div><strong>{s.name}</strong> {s.note ? <span className="text-slate-500">— {s.note}</span> : null}</div>
                        <div className={`font-semibold ${s.side==='buy'? 'text-green-600' : s.side==='sell' ? 'text-rose-600' : 'text-slate-600'}`}>{s.side.toUpperCase()}</div>
                      </div>
                      <div className="text-xs text-slate-500 mt-1">weight: {s.weight}</div>
                    </li>
                  ))}
                </ul>
              </div>

              <div className="mt-3 text-xs text-slate-500">
                <strong>How recommendation is formed:</strong> We compute classic indicators (MA/EMA/RSI/MACD) and assign weights to each signal. The final recommendation is an aggregate weighted score. This is educational and NOT financial advice.
              </div>
            </div>
          )}
        </div>

        <div className="bg-white p-4 rounded shadow-sm mt-4">
          <h4 className="font-medium">Export / Copy</h4>
          <div className="flex gap-2 mt-2">
            <button onClick={()=>{ if(!rawRows || rawRows.length===0) return alert('No data'); const blob = new Blob([JSON.stringify({rows: rawRows, decision, indicators}, null, 2)], {type:'application/json'}); const url = URL.createObjectURL(blob); const a = document.createElement('a'); a.href = url; a.download = `${symbol||'data'}-analysis.json`; a.click(); URL.revokeObjectURL(url); }} className="px-3 py-2 bg-slate-100 rounded">Download JSON</button>
            <button onClick={()=>{ navigator.clipboard?.writeText(JSON.stringify({decision, signals: decision?.signals || []}, null, 2)); alert('Copied short result to clipboard'); }} className="px-3 py-2 bg-slate-100 rounded">Copy result</button>
          </div>
        </div>

      </div>
    </div>

    <div className="mt-6 text-xs text-slate-500">
      <strong>Notes & disclaimers:</strong>
      <ul className="list-disc ml-5 mt-2">
        <li>This tool performs technical analysis on historical price series. It is educational and not financial advice.</li>
        <li>Indicators work best with clean, continuous data. Be mindful of corporate actions, splits, and missing rows.</li>
        <li>For live trading, connect to a broker API and add proper risk management (position sizing, stop-losses).</li>
      </ul>
    </div>

  </div>
</div>

); }
